lib/HLSL/DxilValidation.cpp

```diff
@@ -3988,6 +3988,114 @@ static void ValidateMetadata(ValidationContext &ValCtx) {
+        // would store entry properties
+        const llvm::MDOperand &mOp =
+            TANode->getOperand(TANode->getNumOperands() - 1);
+        // the final operand to the entry points tuple should be a constant int.
+        if (mOp == nullptr ||
+            (mOp.get())->getMetadataID() != Metadata::MDTupleKind) {
+          continue;
+        }
+
+        // get access to the node that stores entry properties
+        MDTuple *TANode2 =
+            dyn_cast<MDTuple>(TANode->getOperand(TANode->getNumOperands() - 1));
+        // find any incompatible tags inside the entry properties
+        // increment j by 2 to only analyze tags, not values
+        bool foundTag = false;
+        for (unsigned j = 0, end2 = TANode2->getNumOperands(); j < end2;
+             j += 2) {
+          // if this tag is already present in the current entry point, emit an
+          // error
+
+          const MDOperand &mOp2 = TANode2->getOperand(j);
+          // note, we are only looking for tags, which will be a constant
+          // integer
+          if (mOp2 == nullptr || (mOp2.get())->getMetadataID() !=
+                                     Metadata::ConstantAsMetadataKind) {
+            continue;
+          }
+          ConstantInt *tag = mdconst::extract<ConstantInt>(mOp2);
+          uint64_t tagValue = tag->getZExtValue();
+
+          // legacy wavesize is only supported between 6.6 and 6.7, so we
+          // should fail if we find the ranged wave size metadata tag
+          if (tagValue == DxilMDHelper::kDxilRangedWaveSizeTag) {
+            if (foundTag) {
+              ValCtx.EmitFormatError(ValidationRule::SmWaveSizeTagDuplicate,
+                                     {});
+              return;
+            }
+            foundTag = true;
+            if (SM->IsSM66Plus() && !SM->IsSM68Plus()) {
+
+              ValCtx.EmitFormatError(
+                  ValidationRule::SmWaveSizeRangeNeedsSM68Plus, {});
+              return;
+            }
+            // get the metadata that contains the parameters to the wavesize
+            // attribute
+            MDTuple *TANode3 = dyn_cast<MDTuple>(TANode2->getOperand(j + 1));
+            if (TANode3->getNumOperands() != 3) {
+              ValCtx.EmitFormatError(
+                  ValidationRule::SmWaveSizeRangeExpectsThreeParams, {});
+              return;
+            }
+            for (int k = 0; k < 3; k++) {
+              const MDOperand &param = TANode3->getOperand(k);
+              if (param->getMetadataID() != Metadata::ConstantAsMetadataKind) {
+                ValCtx.EmitFormatError(
+                    ValidationRule::SmWaveSizeNeedsConstantOperands, {});
+                return;
+              }
+            }
+
+          } else if (tagValue == DxilMDHelper::kDxilWaveSizeTag) {
+            // if the shader model is anything but 6.6 or 6.7, then we do not
+            // expect to encounter the legacy wave size tag.
```

Comment:
I think this comment would be more useful above the conditional that checks for this.